### PR TITLE
Ffmpeg

### PR DESCRIPTION
**Pull request by Bobini1**
This PR includes:
- adding a sound class for decoding any type of sound and playing it
- tests
- factory class
- factory concept
- lua bindings

### CODE REVIEW

File: src/sounds/FFmpegOpenAlSound.cpp [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    ALint size{};
+    alGetBufferi(*sampleBuffer, AL_SIZE, &size);
+    ALint frequency{};
+    alGetBufferi(*sampleBuffer, AL_FREQUENCY, &frequency);
+    ALint channels{};
+    alGetBufferi(*sampleBuffer, AL_CHANNELS, &channels);
+    ALint bits{};
+    alGetBufferi(*sampleBuffer, AL_BITS, &bits);
+    return std::chrono::nanoseconds(static_cast<std::chrono::nanoseconds::rep>(
+      size * 8 / (frequency * channels * bits)));
```
Comment (LukerMaster) [inline src/sounds/FFmpegOpenAlSound.cpp:306]: `size * 8`? 8 of what?

**Review by LukerMaster — COMMENTED** [review]
I hope it helps, I could not find anything super suspicious

File: src/sounds/FFmpegOpenAlSound.cpp (hunk 1 above)
Comment (Bobini1) [reply src/sounds/FFmpegOpenAlSound.cpp:306]: Bits. I'll make a variable for that.

File: src/sounds/FFmpegOpenAlSound.cpp [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    spdlog::info("Sample format: {}", av_get_sample_fmt_name(format));
+    spdlog::info("Sample rate: {}", codecContext->sample_rate);
+    spdlog::info("Channels: {}", codecContext->channels);
+
+    auto samples = decodeFile(*formatContext, *audioStream, *codecContext);
+    alcMakeContextCurrent(getALContext());
+    // load samples into openal
+    ALuint buffer{};
+    alGenBuffers(1, &buffer);
+    sampleBuffer = std::shared_ptr<ALuint>(new ALuint(buffer), deleteBuffer);
```
[the remark this answers is not in the document]
Comment (Bobini1) [reply src/sounds/FFmpegOpenAlSound.cpp:188]: I had to use a custom deleter and `make_shared` doesn't support that.

The deal here is as follows: I would like to be able to have a RAII type which would release the resource properly at destruction. `shared_ptr` supports that. The disadvantage is that it uses the heap to store the int. I'm going to think about it more. Maybe I'll write some `shared_val` class.

File: src/sounds/FFmpegOpenAlSound.cpp [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    if (avformat_open_input(&formatContext, filename, nullptr, nullptr)) {
+        throw std::runtime_error("Could not open file " +
+                                 std::string(filename));
+    }
+
+    if (avformat_find_stream_info(formatContext, nullptr) < 0) {
+        throw std::runtime_error("Could not find stream info");
+    }
+    const auto audioStreamIndex = av_find_best_stream(
+      formatContext, AVMEDIA_TYPE_AUDIO, -1, -1, nullptr, 0);
```
[the remark this answers is not in the document]
Comment (Bobini1) [reply src/sounds/FFmpegOpenAlSound.cpp:160]: Good idea. I honestly don't even know what they mean, I copied this from the internet.

File: src/sounds/FFmpegOpenAlSound.cpp [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                    return AL_FORMAT_STEREO16;
+                case AV_SAMPLE_FMT_FLT:
+                    return AL_FORMAT_STEREO_FLOAT32;
+                case AV_SAMPLE_FMT_DBL:
+                    return AL_FORMAT_STEREO_DOUBLE_EXT;
+                default:
+                    throw std::runtime_error("Unsupported format");
+            }
+        default:
+            throw std::runtime_error("Unsupported format");
```
[the remark this answers is not in the document]
Comment (Bobini1) [reply src/sounds/FFmpegOpenAlSound.cpp:50]: I did it to silence warnings but I agree that it looks bad. I'll try to improve it.

File: src/drawing/SplashScene.h [hunk 5]
```diff
@@ -29,8 +29,8 @@ class SplashScene : public Scene
     bool initialized = false;
     sol::state state;
     AnimationPlayerType animationPlayer;
-    std::shared_ptr<TextureLoaderType> textureLoader;
-    std::shared_ptr<FontLoaderType> fontLoader;
+    TextureLoaderType* textureLoader;
```
[the remark this answers is not in the document]
Comment (Bobini1) [reply src/drawing/SplashScene.h:32]: Naah, it's actually a deeper error. It was always supposed to be a raw pointer. But I didn't change the definition in the class back then and it was wrapping a raw pointer it did not own in a shared_ptr. This lead to a double free and a segfault.

**Pull request merged**